### ISSUE TITLE
[networkd] Reconfiguration of socketcan interface requires networkd restart

### ISSUE THREAD

**Issue by splatch** [systemd/systemd#20434]
**systemd version 248 (248-5-arch)**

This is an branch issue of #20428, it is related to reloading of `systemd-networkd` configurations to change CAN interface settings.
My intention here is to make sure that it is sufficient to use `networkctl reload` and `networkctl reconfigure can0` as promised in documentation.

**Used distribution**
 > `Manjaro / Arch`

**Linux kernel version used** (`uname -a`)
 > `Linux arch 5.13.9-1-MANJARO #1 SMP PREEMPT Sun Aug 8 09:02:03 UTC 2021 x86_64 GNU/Linux`,
 > With downgraded systemd.

**CPU architecture issue was seen on**
 > `x86_64`

**Expected behaviour you didn't see**
 > `Reconfiguration of socketcan interface`

**Unexpected behaviour you saw**
 > `The .network file change is not reflected in interface settings`

**Steps to reproduce the problem**
 > 1. Get a gs_usb dongle
 > 2. Create basic .network configuration as attached
 > 3. Bring interface up
 > 4. Change .network file
 > 5. Reload configurations using `networkctl reload`
 > 6. Reconfigure interface using ` networkctl reconfigure can0`

**Additional program output to the terminal or log subsystem illustrating the issue**

Config file `/etc/systemd/network/80-can0.network`:
```text
[Match]
Name=*can0

[CAN]
BitRate=50000
RestartSec=100ms
```

Debug output from networkd
```text
can0: Flags change: -UP -LOWER_UP -RUNNING
can0: Link DOWN
can0: Lost carrier
can0: State is configuring, dropping config
can0: Configuring CAN link.
can0: Setting bitrate = 50000 bit/s
can0: Setting sample point = 87.5%
can0: Setting restart = 100ms
can0: Bringing CAN link up
can0: Set link
can0: Flags change: +UP +LOWER_UP +RUNNING
can0: Link UP
can0: Gained carrier
can0: link_check_ready(): static addresses are not configured.
can0: link_check_ready(): static addresses are not configured.
Setting '/proc/sys/net/ipv6/conf/can0/proxy_ndp' to '0'
can0: Cannot configure proxy NDP for the interface, ignoring: No such file or directory
can0: State changed: configuring -> configured
// change configuration in .network file
// reload and then reconfigure can0
can0: Flags change: -UP -LOWER_UP -RUNNING
can0: Link DOWN
can0: Lost carrier
// up can0
can0: State is configured, dropping config
can0: Flags change: +UP +LOWER_UP +RUNNING
can0: Link UP
can0: Gained carrier
can0: State changed: configured -> configuring
can0: link_check_ready(): static addresses are not configured.
can0: link_check_ready(): static addresses are not configured.
Setting '/proc/sys/net/ipv6/conf/can0/proxy_ndp' to '0'
can0: Cannot configure proxy NDP for the interface, ignoring: No such file or directory
can0: State changed: configuring -> configured
```

As shown above .network file change, even if it was reloaded is not reflected in interface configuration. Additionally "reconfigure" operation does not bring interface to its previous state.
While I am not certain if later is common I believe that lack of reaction for bitrate change is a bug.

**Comment by yuwata** [comment]
Is this issue reproducible with v249+#20431?

**Comment by splatch** [comment]
@yuwata I have to build systemd from git to confirm, please give me some time to do so.

**Comment by splatch** [comment]
Tested with git version (`systemd 249 (249.r405.g8908ceb7a8-1-arch)`), with 8908ceb7a855eabf5739dc15932ee1d41a0bc93c as a last commit. The result is the same, bitrate does not change when reload and reconfigure sequence is made. I did tweak configuration with `ActivationPolicy=always-up` to make sure that reconfigure does not leave interface down.

Detailed log:
```
can0: Permanent MAC address not found for new device, continuing without: Operation not supported
can0: Link 26 added
can0: Saved original MTU 16 (min: 0, max: 4294967295)
can0: Flags change: +NOARP +ECHO
can0: link pending udev initialization...
can0: udev initialized link
can0: State changed: pending -> initialized
can0: Link state is up-to-date
can0: found matching network '/etc/systemd/network/80-can0.network'.
can0: State changed: initialized -> configuring
can0: Requested to set CAN interface configurations
can0: Requested to activate link
can0: Setting CAN interface configurations
can0: Setting bitrate = 50000 bit/s
can0: Setting sample point = 87.5%
can0: Setting restart = 100ms
can0: CAN interface configurations set.
can0: link_check_ready(): link is not activated.
can0: Bringing link up
can0: Flags change: +UP +LOWER_UP
can0: Link UP
can0: State changed: configuring -> configured
can0: Flags change: +RUNNING
can0: Gained carrier
// change .network, reload, reconfigure can0
can0: Flags change: -UP -LOWER_UP -RUNNING
can0: Link DOWN
can0: ActivationPolicy is "always-on", forcing link up.
can0: Requested to bring link up
can0: Lost carrier
can0: Bringing link up
can0: Flags change: +UP +LOWER_UP +RUNNING
can0: Link UP
can0: Gained carrier
```

**Comment by yuwata** [comment]
Could you paste full debug log at somewhere?

**Comment by splatch** [comment]
@yuwata sure, will output of `SYSTEMD_LOG_LEVEL=debug ./systemd-networkd` will serve you well or shall I add more options?

**Comment by yuwata** [comment]
That should be enough. Thank you.

**Comment by splatch** [comment]
This is output I got https://gist.github.com/splatch/89598e63c939eabb2cfb31baf3aecf28. I will restart system to ensure that everything comes from most recent version and repeat.

**Comment by splatch** [comment]
After restart, with most recent version it works as expected. Thank you very much. I thought I spot a startup issue with v249 but it was a race condition with system networkd. After stopping system unit, so there is only one networkd daemon it works without any issues.
I updated above gist with v249 logs where you can confirm behavior. I am closing this issue since your fix introduced in #20431 covers that as well.
